File: breadth-first-search/breadth-first-search.rb
require "set"

class Node
  attr_reader :neighbors, :number

  def initialize(number)
    @neighbors = Set.new
    @number = number
  end
end

class Graph
  # `node_one` and `node_two` are now neighbors
  # they are both `Node` instances, so they should both
  # have one another in their @neighbors sets
  # since this is an undirected, possibly cyclic graph
  def edge_between(node_one, node_two)
    node_one.neighbors << node_two
    node_two.neighbors << node_one
  end
end

# Breadth-first search (BFS) on an undirected graph
# input: graph and source vertex, target optional
# output: (1) all vertices reachable from source vertex are 'discovered'
# (2) `meta` generation creates a list of 'actions' from start to end. (3) if a
# target is supplied, print the shortest path from start to end as found by BFS
# pseudocode pulled from wiki, visualgo, Skiena's lectures

def breadth_first_search(graph, source, target = nil)
  # queue: first-in, first-out (FIFO). keeps track of which vertices have already
  # been visited but have not yet been visited from, so we know where to search.
  queue = Queue.new

  # a `Set` instance is a collection of unordered values with no duplicates
  # maintain a list of visited nodes and prevent checking a node more than once
  visited = Set.new

  # shortest path information, if applicable
  meta = {}

  # enqueue the source key (push it to the empty queue)
  queue.enq(source)

  until queue.empty?
    # current node, which we `shift` from the queue
    current = queue.deq

    # print the shortest path if it was found
    return path(source, current, meta) if target && current == target

    # we don't have to keep track of distance here, since we have a method
    # to access each of the node's neighbors. the neighbors are stored in a set.
    # process each neighboring vertex of the current node,
    # i.e. traverse all outgoing edges from the current node.
    current.neighbors.each do |neighbor|
      # if the neighbor node is unvisited, we ignore this edge
      unless visited.include?(neighbor)
        queue.enq(neighbor)
        visited.add(neighbor) # we just enqueued this node, so mark it as visited
        meta[neighbor] = current # record the path (only done once, b/c of `unless`)
      end
    end
  end
end

def path(source, target, meta)
  # retrieve the path to the target node, using the meta hash
  # that was constructed at each iteration of the `each` iteration
  predecessor = target
  result_path = []

  # follow the leads from the back
  until predecessor == source
    result_path << predecessor.number
    predecessor = meta[predecessor]
  end

  # add the source, since now `predecessor == source` is true
  result_path << source.number
  result_path.reverse.join(", ")
end

# example

n0 = Node.new("0")
n1 = Node.new("1")
n2 = Node.new("2")
n3 = Node.new("3")
n4 = Node.new("4")
n5 = Node.new("5")
n6 = Node.new("6")
n7 = Node.new("7")

graph1 = Graph.new
graph1.edge_between(n0, n1)
graph1.edge_between(n1, n2)
graph1.edge_between(n1, n3)
graph1.edge_between(n2, n3)
graph1.edge_between(n3, n4)
graph1.edge_between(n4, n5)
graph1.edge_between(n4, n6)
graph1.edge_between(n5, n7)
graph1.edge_between(n6, n7)
#
# 0 - 1 - 3 - 4 - 5
#     |  /    |   |
#     | /     |   |
#     2       6 - 7
#
puts breadth_first_search(graph1, n1, n4) # => 1, 3, 4
# q: 1
# enter loop -> queue is not empty
# dequeue 1. q: []
# current node -> 1
# visit all neighbors/children: expecting 0, 2 and 3
# enqueue 0, 2, 3, marking each as visited. q: [0, 2, 3]
# deq 0. q: [2, 3]
# current node -> 0
# visit all neighbors/children: expecting 1
# enqueue 1, marking as visited. q: [2, 3, 1]
# deq 2. q: [3, 1]
# current node -> 2
# visit all neighbors/children: they're already visit
# deq 3. q: [1]
# current node -> 3
# visit all neighbors/children: expecting 4
# enqueue 4, marking as visited. q: [1, 4]
# deq 1. q: [4]
# we are at the target, so break & construct path
# the last queue is [4]
# the path is then essentially constructed from { 4 => 3, 3 => 1 }

puts breadth_first_search(graph1, n3, n6) # => 3, 4, 6

# time complexity
# with an adjacency list, O(|V| + |E|) because
# we inspect _every_ vertex _once_ (due to visited set) -> |V|
# we explore `k` neighbors for vertex v
## if all vertices are explored, we examine all edges
## & total neighbors of all vertices -> |E|
# TODO: adjacency list

File: breadth-first-search/breadth-first-search.js
// Breadth-first search (BFS)

// - - - - - - - - - - - - - - - -
// (4) with ES6 classes & constructors syntax
// - - - - - - - - - - - - - - - -

class Graph {
  constructor(nodes = []) {
    this.nodes = nodes;
    this.neighbors = new Array(nodes).fill().map(() => []);
    this.visited = new Array(nodes).fill().map(() => false);
    this.parents = new Array(nodes).fill().map(() => null);
  }

  edgeBetween(nodeOne, nodeTwo) {
    // create an adjacency list to record edges between nodes
    this.neighbors[nodeOne].push(nodeTwo);
    this.neighbors[nodeTwo].push(nodeOne);
  }
};

class BreadthFirstSearch {
  constructor(graph, source, target) {
    this.source = source;
    this.target = target;
    this.graph = graph;
  }

  search() {
    const graph = this.graph;
    const source = this.source;
    const target = this.target;

    let queue = [source];

    while (queue.length) {
      let currentNode = queue.shift();

      if (currentNode == target) return this.path(currentNode);

      let neighbors = graph.neighbors[currentNode];
      for (let i = 0; i < neighbors.length; i += 1) {
        let neighbor = neighbors[i];
        if (!graph.visited[neighbor]) {
          queue.push(neighbor);
          graph.visited[neighbor] = true;
          graph.parents[neighbor] = currentNode;
        }
      }
    }
  }

  path(currentNode) {
    let path = [];
    let predecessor = currentNode;

    while (predecessor !== this.source) {
      path.push(predecessor);
      predecessor = this.graph.parents[predecessor];
    }

    return path.concat(predecessor).reverse();
  }
};

const graphOne = new Graph(8); // 7 nodes

graphOne.edgeBetween(0, 1);
graphOne.edgeBetween(1, 2);
graphOne.edgeBetween(1, 3);
graphOne.edgeBetween(2, 3);
graphOne.edgeBetween(3, 4);
graphOne.edgeBetween(4, 5);
graphOne.edgeBetween(4, 6);
graphOne.edgeBetween(5, 7);
graphOne.edgeBetween(6, 7);

// 0 - 1 - 3 - 4 - 5
//     |  /    |   |
//     | /     |   |
//     2       6 - 7

const bfsOne = new BreadthFirstSearch(graphOne, 1, 4);
console.log(bfsOne.search());

const graphTwo = new Graph(8); // 7 nodes

graphTwo.edgeBetween(0, 1);
graphTwo.edgeBetween(1, 2);
graphTwo.edgeBetween(1, 3);
graphTwo.edgeBetween(2, 3);
graphTwo.edgeBetween(3, 4);
graphTwo.edgeBetween(4, 5);
graphTwo.edgeBetween(4, 6);
graphTwo.edgeBetween(5, 7);
graphTwo.edgeBetween(6, 7);

// 0 - 1 - 3 - 4 - 5
//     |  /    |   |
//     | /     |   |
//     2       6 - 7

const bfsTwo = new BreadthFirstSearch(graphTwo, 3, 6);
console.log(bfsTwo.search());

// - - - - - - - - - - - - - - - -
// (3) implementation via objects/lists input
// - - - - - - - - - - - - - - - -

// TODO

// - - - - - - - - - - - - - - - -
// (2) implementation via adjacency matrix input
// - - - - - - - - - - - - - - - -

const path = function path(meta, source, target) {
  let predecessor = target
  let result_path = [];

  while (predecessor !== source) {
    result_path.push(predecessor);
    predecessor = meta[predecessor];

  }

  result_path.push(source);
  return result_path.reverse().join(", ");
};

const breadthFirstSearch2 = function breadthFirstSearch(graph, source, target) {
  let meta = []; // to print our output path to the target
  let visited = [];
  let queue = [source];

  while (queue.length) {
    currentNode = queue.shift();
    if (currentNode === target) return path(meta, source, target);

    for (let i = 0; i < graph.length; i += 1) {
      let neighbor = graph[currentNode][i];
      // if this node hasn't been visited before &&
      // this node is a neighbor (adjacent to) the currentNode (value 1),
      // then mark it as visited && build the `meta` array for output
      if (!visited[i] && graph[currentNode][i] && i !== currentNode) {
        meta[i] = currentNode;
        visited[i] = true;
        queue.push(i);
      }
    }
  }

  return "No path found!";
}

console.log("\nAdjacency Matrix");

// https://visualgo.net/en/graphds
const testBFS2 = [
  [0, 1, 0, 0, 0, 0, 0, 1, 0, 0],
  [1, 0, 1, 1, 0, 0, 1, 0, 0, 0],
  [0, 1, 0, 0, 0, 0, 0, 0, 0, 0],
  [0, 1, 0, 0, 1, 1, 0, 0, 0, 0],
  [0, 0, 0, 1, 0, 0, 0, 0, 0, 0],
  [0, 0, 0, 1, 0, 0, 0, 0, 0, 0],
  [0, 1, 0, 0, 0, 0, 0, 0, 0, 0],
  [1, 0, 0, 0, 0, 0, 0, 0, 1, 1],
  [0, 0, 0, 0, 0, 0, 0, 1, 0, 0],
  [0, 0, 0, 0, 0, 0, 0, 1, 0, 0]
];

const performSearch = breadthFirstSearch2(testBFS2, 0, 9); // 0, 7, 9
const performSearch2 = breadthFirstSearch2(testBFS2, 6, 5); // 6, 1, 3, 5
console.log(performSearch);
console.log(performSearch2);
console.log('\n');

// - - - - - - - - - - - - - - - -
// (1) implementation via adjacency list & Queue abstraction from KhanAcademy
// - - - - - - - - - - - - - - - -

// Implement a Queue with JS built-in functions
const Queue = function Queue() {
  // Our queue is just an array
  // JS already has built-ins for dequeue (shift) and enqueue (push)
  this.items = [];
}

Queue.prototype.enqueue = function(object) {
  // Side effect
  this.items.push(object);
};

Queue.prototype.dequeue = function() {
  return this.items.shift();
};

Queue.prototype.isEmpty = function() {
  return this.items.length === 0;
};

const doBFS = function(graph, source) {
  let bfsInfo = [];

  // initialize everything to null
  for (let i = 0; i < graph.length; i += 1) {
    bfsInfo[i] = {
      distance: null,
      predecessor: null
    };
  }

  // assign source a distance of 0
  bfsInfo[source].distance = 0;

  let queue = new Queue();

  // begin queue with source node
  queue.enqueue(source);

  while (!queue.isEmpty()) {
    let currentNode = queue.dequeue();

    // in an adjacency list (arr of arrs), the index represents the node,
    // thus we can access the neighbors of a node by the index, and then
    // we can individually access each neighbor in the array
    for (let j = 0; j < graph[currentNode].length; j += 1) {
      let neighbor = graph[currentNode][j];

      // i.e. if it's unvisited (the distance is `null` if so)
      if (bfsInfo[neighbor].distance === null) {
        bfsInfo[neighbor].distance = bfsInfo[currentNode].distance + 1;
        bfsInfo[neighbor].predecessor = currentNode;
        queue.enqueue(neighbor);
      }
    }
  }

  return bfsInfo;
}

const adjList = [
    [1],
    [0, 4, 5],
    [3, 4, 5],
    [2, 6],
    [1, 2],
    [1, 2, 6],
    [3, 5],
    []
    ];

const bfsInfo = doBFS(adjList, 3);

console.log("KhanAcademy BFS with Adjacency List");
for (let i = 0; i < adjList.length; i++) {
    console.log(("vertex " + i + ": distance = " + bfsInfo[i].distance + ", predecessor = " + bfsInfo[i].predecessor));
}

// time complexity
// with an adjacency list, O(|V| + |E|) because
// we inspect _every_ vertex _once_ (due to visited set) -> |V|
// we explore `k` neighbors for vertex v
// if all vertices are explored, we examine all edges
// & total neighbors of all vertices -> |E|
